src/util/macro_util.rs

```diff
@@ -285,6 +285,32 @@ mod size_to_tag {
 #[doc(hidden)]
 pub type SizeToTag<const SIZE: usize> = <() as size_to_tag::SizeToTag<SIZE>>::Tag;
 
+// We put `Sized` in its own module so it can have the same name as the standard
+// library `Sized` without shadowing it in the parent module.
+#[cfg(zerocopy_diagnostic_on_unimplemented)]
+mod __size_of {
+    #[diagnostic::on_unimplemented(
+        message = "`{Self}` is unsized",
+        label = "`IntoBytes` needs all field types to be `Sized` in order to determine whether there is inter-field padding",
+        note = "consider using `#[repr(packed)]` to remove inter-field padding",
+        note = "`IntoBytes` does not require the fields of `#[repr(packed)]` types to be `Sized`"
+    )]
+    pub trait Sized: core::marker::Sized {}
+    impl<T: core::marker::Sized> Sized for T {}
+
+    #[inline(always)]
+    #[must_use]
+    #[allow(clippy::needless_maybe_sized)]
+    pub const fn size_of<T: Sized + ?core::marker::Sized>() -> usize {
```

Comment:
This is such a cursed bound.

Reply:
I just stumbled across this randomly: why is that bound there? Wouldn't any `T` automatically also be `core::mem::Sized` due to the trait bound of `__size_of::Sized`? Thanks!

Reply:
From a correctness perspective, you're right. However, without `?core::marker::Sized`, when Rust encounters a type that is unsized that is passed to this `size_of`, it emits a `Sized` error first and bails, so it never gets to the point of discovering that the `Sized` bound (our bespoke `Sized` trait) isn't satisfied, and so it never emits our custom diagnostic. Adding `?core::marker::Sized` "uncovers" our unsatisfied trait bound so that the desired diagnostic actually gets emitted.

It is, as all things in proc macro land, mildly cursed 😛

Reply:
@joshlf thanks for the explanation!

Reply:
np!